go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -0,0 +1,1768 @@
+	// delete the NB DB set up correctly when searching the DB based on the
+	// name.
+	name := ""
+
+	// Initialize a status which will be used to compare against
+	// new.spec.egressIPs and decide on what from the status should get deleted
+	// or kept.
+	status := []egressipv1.EgressIPStatusItem{}
+
+	// Initialize two empty objects as to avoid SIGSEGV. The code should play
+	// nicely with empty objects though.
+	newEIP := &egressipv1.EgressIP{}
+
+	// Initialize a sets.String which holds egress IPs that were not fully assigned
+	// but are allocated and they are meant to be removed.
+	staleEgressIPs := sets.NewString()
+	if old != nil {
+		name = old.Name
+		status = old.Status.Items
+		staleEgressIPs.Insert(old.Spec.EgressIPs...)
+	}
+	if new != nil {
+		newEIP = new
+		name = newEIP.Name
+		status = newEIP.Status.Items
+		if staleEgressIPs.Len() > 0 {
+			for _, egressIP := range newEIP.Spec.EgressIPs {
+				if staleEgressIPs.Has(egressIP) {
+					staleEgressIPs.Delete(egressIP)
+				}
+			}
+		}
+	}
+
+	// Validate the spec and use only the valid egress IPs when performing any
+	// successive operations, theoretically: the user could specify invalid IP
+	// addresses, which would break us.
+	validSpecIPs, err := eIPC.validateEgressIPSpec(name, newEIP.Spec.EgressIPs)
+	if err != nil {
+		return fmt.Errorf("invalid EgressIP spec, err: %v", err)
+	}
+
+	// Validate the status, on restart it could be the case that what might have
+	// been assigned when ovnkube-master last ran is not a valid assignment
+	// anymore (specifically if ovnkube-master has been crashing for a while).
+	// Any invalid status at this point in time needs to be removed and assigned
+	// to a valid node.
+	validStatus, invalidStatus := eIPC.validateEgressIPStatus(name, status)
+	for status := range validStatus {
+		// If the spec has changed and an egress IP has been removed by the
+		// user: we need to un-assign that egress IP
+		if !validSpecIPs.Has(status.EgressIP) {
+			invalidStatus[status] = ""
+			delete(validStatus, status)
+		}
+	}
+
+	invalidStatusLen := len(invalidStatus)
+	if invalidStatusLen > 0 {
+		metrics.RecordEgressIPRebalance(invalidStatusLen)
+	}
+
+	// Add only the diff between what is requested and valid and that which
+	// isn't already assigned.
+	ipsToAssign := validSpecIPs
+	ipsToRemove := sets.New[string]()
+	statusToAdd := make([]egressipv1.EgressIPStatusItem, 0, len(ipsToAssign))
+	statusToKeep := make([]egressipv1.EgressIPStatusItem, 0, len(validStatus))
+	for status := range validStatus {
+		statusToKeep = append(statusToKeep, status)
+		ipsToAssign.Delete(status.EgressIP)
+	}
+	statusToRemove := make([]egressipv1.EgressIPStatusItem, 0, invalidStatusLen)
+	for status := range invalidStatus {
+		statusToRemove = append(statusToRemove, status)
+		ipsToRemove.Insert(status.EgressIP)
+	}
+	if ipsToRemove.Len() > 0 {
+		// The following is added as to ensure that we only add after having
```

Comment:
I know this comment is copied but please rephrase it so it reflects the new behavior. egressIPController doesn't configure the DBs.

Reply:
I think it is still relevant to keep the OVN DB parts, just that we need to remember this will be done asynchronously.. I will clarify this up.